### PR TITLE
[AIRFLOW-492] Make sure stat updates cannot fail a task

### PR DESCRIPTION
**Pull request by bolkedebruin**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-492

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

Previously a failed commit into the db for the statistics
could also fail a task. Secondly, the ui could display
out of date statistics.

This patch reworks DagStat so that failure to update the
statistics does not propagate. Next to that, it
make sure the ui always displays the latest statistics.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- more comprehensive unit tests

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by bolkedebruin** [comment]
cc @criccomini you might want to include it in 1.8.1 if it passes (as tasks could fail due to duplicate entries), but 1.8.2 is also fine :-)

**Comment by bolkedebruin** [comment]
@aoen @saguziel ping

**Comment by aoen** [comment]
LGTM

**Comment by bolkedebruin** [comment]
cc @criccomini maybe for 1.8.1?

File: airflow/models.py [hunk 1]
```diff
@@ -3857,48 +3857,104 @@ def __init__(self, dag_id, state, count, dirty=False):
...
-        if not dirty_ids:
-            return
+            ids = set([dag_stat.dag_id for dag_stat in qry])
 
-        qry = (
-            session.query(DagRun.dag_id, DagRun.state, func.count('*'))
-            .filter(DagRun.dag_id.in_(dirty_ids))
-            .group_by(DagRun.dag_id, DagRun.state)
-        )
+            # avoid querying with an empty IN clause
```
Comment (saguziel) [inline airflow/models.py:3908]: @bolkedebruin I was debugging the high db load in 1.8.0, and I found that the query for the empty in clause caused most of our load, so we went from 95% to 15% load after putting a guard against this query. Nice foresight! (it seems MySQL can't use the optimization to mark a.b != a.b as impossible if b is nullable)

Comment (bolkedebruin) [reply airflow/models.py:3908]: @saguziel Umpf that sounds pretty bad as those empty in clauses are literally everywhere. @skudriashev did some cleaning up. We don't see it as we a running on Postgres that does not suffer from the issue. Still avoiding round trips to the db is good practice.

Btw if you are looking at db load you want to review (pls :) ) this one as well: https://github.com/apache/incubator-airflow/pull/2267

**Pull request closed without merging**